File: scripts/migrate-to-firebase.js
const { initializeApp, cert } = require('firebase-admin/app');
const { getFirestore } = require('firebase-admin/firestore');
const Database = require('better-sqlite3');
const path = require('path');

// Firebase Admin 초기화 (서비스 계정 키 사용)
const serviceAccount = require('./firebase-admin-key.json');

initializeApp({
  credential: cert(serviceAccount)
});

const firestore = getFirestore();

// SQLite DB 연결
const DB_PATH = '/Users/kangkukjin/Desktop/AI/indiebizOS/data/packages/installed/tools/blog/data/blog_insight.db';
const sqliteDb = new Database(DB_PATH, { readonly: true });

// 숨길 카테고리
const HIDDEN_CATEGORIES = ['임시보관함', '집자료들'];
const CATEGORY_REMAP = { '재검토 글들': '미분류' };

function shouldHide(category) {
  if (!category) return false;
  return HIDDEN_CATEGORIES.some(hidden => category.startsWith(hidden));
}

function remapCategory(category) {
  if (!category) return category;
  const mainCat = category.includes('/') ? category.split('/')[0] : category;
  if (CATEGORY_REMAP[mainCat]) {
    return category.replace(mainCat, CATEGORY_REMAP[mainCat]);
  }
  return category;
}

async function migrate() {
  console.log('마이그레이션 시작...\n');

  // 1. Posts 마이그레이션
  console.log('1. Posts 테이블 마이그레이션...');
  const posts = sqliteDb.prepare(`
    SELECT id, post_id, title, category, pub_date, content, char_count
    FROM posts
  `).all();

  let postCount = 0;
  let skippedCount = 0;
  const batch1 = firestore.batch();
  const batches = [batch1];
  let currentBatch = batch1;
  let batchCount = 0;

  for (const post of posts) {
    // 숨김 카테고리 필터링
    if (shouldHide(post.category)) {
      skippedCount++;
      continue;
    }

    const docRef = firestore.collection('posts').doc(post.post_id);
    currentBatch.set(docRef, {
      id: post.id,
      post_id: post.post_id,
      title: post.title,
      category: remapCategory(post.category),
      pub_date: post.pub_date,
      content: post.content,
      char_count: post.char_count,
      created_at: new Date()
    });

    postCount++;
    batchCount++;

    // Firestore batch는 500개 제한
    if (batchCount >= 400) {
      const newBatch = firestore.batch();
      batches.push(newBatch);
      currentBatch = newBatch;
      batchCount = 0;
      console.log(`  - ${postCount}개 처리중...`);
    }
  }

  console.log(`  - 총 ${postCount}개 글 준비 완료 (${skippedCount}개 스킵)`);

  // 2. Summaries 마이그레이션
  console.log('\n2. Summaries 테이블 마이그레이션...');
  const summaries = sqliteDb.prepare(`
    SELECT s.post_id, s.summary, s.keywords
    FROM summaries s
    INNER JOIN posts p ON s.post_id = p.post_id
  `).all();

  let summaryCount = 0;
  for (const summary of summaries) {
    // 해당 포스트가 숨김 카테고리인지 확인
    const post = posts.find(p => p.post_id === summary.post_id);
    if (post && shouldHide(post.category)) {
      continue;
    }

    const docRef = firestore.collection('summaries').doc(summary.post_id);
    currentBatch.set(docRef, {
      post_id: summary.post_id,
      summary: summary.summary,
      keywords: summary.keywords
    });

    summaryCount++;
    batchCount++;

    if (batchCount >= 400) {
      const newBatch = firestore.batch();
      batches.push(newBatch);
      currentBatch = newBatch;
      batchCount = 0;
    }
  }

  console.log(`  - 총 ${summaryCount}개 요약 준비 완료`);

  // 3. 배치 커밋
  console.log(`\n3. Firestore에 저장 중... (${batches.length}개 배치)`);
  for (let i = 0; i < batches.length; i++) {
    await batches[i].commit();
    console.log(`  - 배치 ${i + 1}/${batches.length} 완료`);
  }

  console.log('\n마이그레이션 완료!');
  console.log(`- Posts: ${postCount}개`);
  console.log(`- Summaries: ${summaryCount}개`);
}

migrate().catch(console.error);

File: scripts/migrate-to-firebase.py
#!/usr/bin/env python3
"""
SQLite -> Firestore 마이그레이션 스크립트

사용법:
1. pip install firebase-admin
2. firebase-admin-key.json 파일을 이 스크립트와 같은 폴더에 둠
3. python migrate-to-firebase.py
"""
import sqlite3
import os
import firebase_admin
from firebase_admin import credentials, firestore

# 스크립트 위치 기준으로 경로 설정
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))

# Firebase Admin 초기화
cred = credentials.Certificate(os.path.join(SCRIPT_DIR, 'firebase-admin-key.json'))
firebase_admin.initialize_app(cred)

db = firestore.client()

# SQLite DB 경로 (사용자 환경에 맞게 수정)
DB_PATH = '/Users/kangkukjin/Desktop/AI/indiebizOS/data/packages/installed/tools/blog/data/blog_insight.db'

# 숨길 카테고리
HIDDEN_CATEGORIES = ['임시보관함', '집자료들']
CATEGORY_REMAP = {'재검토 글들': '미분류'}

def should_hide(category):
    if not category:
        return False
    return any(category.startswith(hidden) for hidden in HIDDEN_CATEGORIES)

def remap_category(category):
    if not category:
        return category
    main_cat = category.split('/')[0] if '/' in category else category
    if main_cat in CATEGORY_REMAP:
        return category.replace(main_cat, CATEGORY_REMAP[main_cat])
    return category

def migrate():
    print('마이그레이션 시작...\n')

    # SQLite 연결
    sqlite_conn = sqlite3.connect(DB_PATH)
    sqlite_conn.row_factory = sqlite3.Row
    cursor = sqlite_conn.cursor()

    # 1. Posts 마이그레이션
    print('1. Posts 테이블 마이그레이션...')
    cursor.execute('''
        SELECT id, post_id, title, category, pub_date, content, char_count
        FROM posts
    ''')
    posts = cursor.fetchall()

    post_count = 0
    skipped_count = 0
    batch = db.batch()
    batch_count = 0

    posts_dict = {}  # 나중에 summaries 필터링용

    for post in posts:
        post_dict = dict(post)
        posts_dict[post_dict['post_id']] = post_dict

        # 숨김 카테고리 필터링
        if should_hide(post_dict['category']):
            skipped_count += 1
            continue

        doc_ref = db.collection('posts').document(post_dict['post_id'])
        batch.set(doc_ref, {
            'id': post_dict['id'],
            'post_id': post_dict['post_id'],
            'title': post_dict['title'],
            'category': remap_category(post_dict['category']),
            'pub_date': post_dict['pub_date'],
            'content': post_dict['content'],
            'char_count': post_dict['char_count']
        })

        post_count += 1
        batch_count += 1

        # Firestore batch는 500개 제한
        if batch_count >= 400:
            batch.commit()
            batch = db.batch()
            batch_count = 0
            print(f'  - {post_count}개 처리중...')

    # 남은 배치 커밋
    if batch_count > 0:
        batch.commit()

    print(f'  - 총 {post_count}개 글 완료 ({skipped_count}개 스킵)')

    # 2. Summaries 마이그레이션
    print('\n2. Summaries 테이블 마이그레이션...')
    cursor.execute('''
        SELECT post_id, summary, keywords FROM summaries
    ''')
    summaries = cursor.fetchall()

    summary_count = 0
    batch = db.batch()
    batch_count = 0

    for summary in summaries:
        summary_dict = dict(summary)

        # 해당 포스트가 숨김 카테고리인지 확인
        post = posts_dict.get(summary_dict['post_id'])
        if post and should_hide(post['category']):
            continue

        doc_ref = db.collection('summaries').document(summary_dict['post_id'])
        batch.set(doc_ref, {
            'post_id': summary_dict['post_id'],
            'summary': summary_dict['summary'],
            'keywords': summary_dict['keywords']
        })

        summary_count += 1
        batch_count += 1

        if batch_count >= 400:
            batch.commit()
            batch = db.batch()
            batch_count = 0
            print(f'  - {summary_count}개 처리중...')

    # 남은 배치 커밋
    if batch_count > 0:
        batch.commit()

    print(f'  - 총 {summary_count}개 요약 완료')

    print('\n✅ 마이그레이션 완료!')
    print(f'- Posts: {post_count}개')
    print(f'- Summaries: {summary_count}개')

    sqlite_conn.close()

if __name__ == '__main__':
    migrate()
